runtime/lua/vim/shared.lua

```diff
@@ -504,5 +504,111 @@ function vim.is_callable(f)
   return type(m.__call) == 'function'
 end
 
+
+--- Returns an object with methods to be used to log to `filename`
+--
+-- Example:
+--   local log = vim.new_logger('vim-lsp.log')
+--   local _ = log.info() and log.info('Log this message', some_var, another_var)
+--
+-- Loggers are cached by filename.
+--
+--@param filename filename as string
+--@param fmt_f a function used to format the arguments.
+--             Optional, defaults to `vim.inspect(arg, {newline=''})`
+--@return a logger which has the following functions to log:
+--
+--          .trace(...)
+--          .debug(...)
+--          .info(...)
+--          .warn(...)
+--          .error(...)
+--
+--        Each method can be called without arguments to check if the log level
+--        matches. If called with arguments each argument is formatted via
+--        `fmt_f` and logged to `stdpath('data') </> filename`.
+--
+--        It also contains
+--
+--          .levles           -- a table with the available log levels
+--          .set_level(level) -- set the level to one of (`trace`, `debug`,
+--                               `info`, `warn`, `error`)
+--          .get_filename()   -- Returns the full path to the logfile
+--
+--
+local loggers = {}
+function vim.new_logger(filename, fmt_f)
+  local log = loggers[filename]
+  if log then return log end
+  log = {}
+
+  local levels = {
+    TRACE = 0;
+    DEBUG = 1;
+    INFO  = 2;
+    WARN  = 3;
+    ERROR = 4;
+  }
+  local log_date_format = "%FT%H:%M:%SZ%z"
+  local path_sep = vim.loop.os_uname().sysname == "Windows" and "\\" or "/"
+  local function path_join(...)
+    return table.concat(vim.tbl_flatten{...}, path_sep)
+  end
+  local logfilename = path_join(vim.fn.stdpath('data'), filename)
+  log.levels = vim.deepcopy(levels)
+  vim.tbl_add_reverse_lookup(log.levels)
+
+  function log.get_filename()
+    return logfilename
+  end
+
+  local current_log_level = levels.WARN
+  function log.set_level(level)
+    if type(level) == 'string' then
+      current_log_level = assert(log.levels[level:upper()], string.format("Invalid log level: %q", level))
+    else
+      assert(type(level) == 'number', "level must be a number or string")
+      current_log_level = assert(log.levels[level], string.format("Invalid log level: %d", level))
+    end
+  end
+
+  vim.fn.mkdir(vim.fn.stdpath('data'), "p")
+  local logfile = assert(io.open(logfilename, "a+"))
+
+  function log.close()
+    logfile:close()
+    loggers[filename] = nil
+  end
+
+  local fmt = fmt_f or function(x) return vim.inspect(x, {newline=''}) end
```

Comment:
I think this should be saved onto the `log` object itself. That way later you can update the `fmt` function (i.e., override the default fmt function in the logger returned by `vim.lsp.log`).

Does that seem good?

Then you should call below `log.fmt(arg)`

Reply:
@tjdevries makes sense to me - I've pushed a fixup with the change.